viime/views.py

```diff
@@ -724,6 +725,56 @@ def get_pca_overview(validated_table: ValidatedMetaboliteTable):
     return Response(png_content, mimetype='image/png')
 
 
+@csv_bp.route('/csv/<uuid:csv_id>/analyses/plsda', methods=['GET'])
+@use_kwargs({
+    'num_of_components': fields.Integer(required=False)
+})
+@load_validated_csv_file
+def get_plsda(validated_table: ValidatedMetaboliteTable, num_of_components: Optional[int] = None):
+    measurements = validated_table.measurements
+    groups = validated_table.groups
+
+    rows, cols = measurements.shape[:2]
+    max_components = min(rows, cols)
+    if num_of_components is None:
+        num_of_components = max_components
+    if num_of_components > max_components:
+        return jsonify({
+            'error': 'invalid num_of_components'
+        }), 400
+
+    scores = plsda(measurements, groups, num_of_components, 'scores')
+    loadings = plsda(measurements, groups, num_of_components, 'loadings')
+
+    # massage the data returned from opencpu to correct format for client:
+    column_names = list(measurements.columns)
+    formatted_loadings = [{
+        'col': col,
+        'loadings': [
+            # loading for i in range(num_of_components) for loading in loadings.get(f'comp{i+1}')
+        ]
+    } for col in column_names]
+
+    for i in range(num_of_components):
+        for j, loading in enumerate(loadings.get(f'comp{i+1}')):
+            formatted_loadings[j]['loadings'].append(loading)
+
+    x = [[] for i in range(len(measurements))]
+
+    for i in range(num_of_components):
+        for j, score in enumerate(scores.get(f'variates.X.comp{i+1}')):
+            x[j].append(score)
+
+    explained_variances = [
+        scores.get(f'explained_variance.comp.{i+1}') for i in range(num_of_components)
+    ]
```

Comment:
Avoid making two loops to preallocate.  Instead, use `.append`.  You might even group the three loops over `num_of_components` together into a single loop.

Reply:
Is this what you mean by using `.append`? https://github.com/girder/viime/pull/613/commits/ed4451db8b56eb083bfcedbbfb42a1a95de24889